### PR TITLE
sql/catalog: Add support for variadic functions

### LINKED ISSUE #761: Support for json_build_object is broken in 1.5
Because json_build_object's arguments are variable the argument check will not pass unless it is called with no arguments.

### PR DESCRIPTION
**Pull request by kyleconroy**
Fixes #761

### CODE REVIEW

File: internal/sql/catalog/catalog.go [hunk 1]
```diff
@@ -251,6 +251,7 @@ type Argument struct {
 	Name       string
 	Type       *ast.TypeName
 	HasDefault bool
+	Variadic   bool
```
Comment (ovadbar) [inline internal/sql/catalog/catalog.go:254]: Why did you add this struct key when there is a mode for Variadic that you can use to check?

File: internal/engine/postgresql/pg_catalog.go [hunk 2]
```diff
@@ -11646,13 +11646,23 @@ func genPGCatalog() *catalog.Schema {
...
+			},
 			ReturnType: &ast.TypeName{Name: "json"},
 		},
 		{
-			Name:       "json_build_object",
-			Args:       []*catalog.Argument{},
+			Name: "json_build_object",
+			Args: []*catalog.Argument{
+				{
+					Type:     &ast.TypeName{Name: "any"},
```
Comment (ovadbar) [inline internal/engine/postgresql/pg_catalog.go:11662]: Shouldn't you mark Variadic arguments as having defaults because there can b 0 of them otherwise the following would fail.

```
postgres=# select json_build_object();
 json_build_object
-------------------
 {}
(1 row)
```

File: internal/endtoend/testdata/json_build/go/query.sql.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+		&i.JsonBuildArray_2,
+		&i.JsonBuildArray_3,
+		&i.JsonBuildArray_4,
+	)
+	return i, err
+}
+
+const selectJSONBuildObject = `-- name: SelectJSONBuildObject :one
+SELECT
+  json_build_object('foo'),
```
Comment (ovadbar) [inline internal/endtoend/testdata/json_build/go/query.sql.go:40]: I would add the following edge case to the tests this will likely fail.
```
json_build_object(),
```

Also I would remove the following tests because they actually fail in postgresql with the error
"ERROR:  argument list must have even number of elements"
```
  json_build_object('foo'),
  json_build_object('foo', 1, 2),
```
But catching those cases would be a little more difficult to catch without custom coding.
